### PR TITLE
Move ProtectedBrowserStorage to the shared fx

### LINKED ISSUE #25534: Move protected browser storage into the shared framework
Currently the protected browser storage functionality is in a separate package. We think this functionality should live in shared framework to remove the need to add the package and to improve serviceability.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #25534

### CODE REVIEW

File: src/Components/Server/src/DependencyInjection/ComponentServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -33,6 +34,9 @@ public static IServerSideBlazorBuilder AddServerSideBlazor(this IServiceCollecti
 
             services.AddDataProtection();
 
+            services.TryAddScoped<ProtectedLocalStorage>();
+            services.TryAddScoped<ProtectedSessionStorage>();
```
Comment (SteveSandersonMS) [inline src/Components/Server/src/DependencyInjection/ComponentServiceCollectionExtensions.cs:38]: This surprises me a bit. Do we definitely want to include the protected storage services by default?

For non-core functionality like this it seems more idiomatic for the developer to call `services.AddProtectedBrowserStorage()` (or similar) themselves to opt in.

I suppose that is a bit of an arbitrary hoop we're giving them to jump through, but it does mean they are making more of a conscious choice to plug it into their application.

Perhaps this is a question for API review.

File: src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs [hunk 2]
```diff
@@ -8,7 +8,10 @@ internal static class BasicTestAppWebDriverExtensions
     {
         public static IWebElement MountTestComponent<TComponent>(this IWebDriver browser) where TComponent : IComponent
         {
-            var componentTypeName = typeof(TComponent).FullName;
+            var componentType = typeof(TComponent);
+            var componentTypeName = componentType.Assembly == typeof(BasicTestApp.Program).Assembly ?
+                componentType.FullName :
+                componentType.AssemblyQualifiedName;
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs:14]: Interesting. Why is this change needed?

File: src/Components/test/testassets/BasicTestApp/Index.razor [hunk 3]
```diff
@@ -66,8 +66,13 @@
...
         <option value="BasicTestApp.PropertiesChangedHandlerParent">Parent component that changes parameters on child</option>
-        <option value="BasicTestApp.ProtectedBrowserStorageUsageComponent">Protected browser storage usage</option>
-        <option value="BasicTestApp.ProtectedBrowserStorageInjectionComponent">Protected browser storage injection</option>
+        @if (!OperatingSystem.IsBrowser())
+        {
+            var serverAssembly = AppDomain.CurrentDomain.GetAssemblies().First(f => f.GetName().Name == "Components.TestServer");
+
+            <option value="Components.TestServer.ProtectedBrowserStorageUsageComponent, @serverAssembly.FullName">Protected browser storage usage</option>
+            <option value="Components.TestServer.ProtectedBrowserStorageInjectionComponent, @serverAssembly.FullName">Protected browser storage injection</option>
+        }
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/BasicTestApp/Index.razor:75]: I'm sure this is viable as-is, but if you feel like you have time to tweak it a bit more, I'd suggest it would be preferable not to have the `@if` statement, because then it would be really confusing if you were in the browser UI looking for something in the list that wasn't there because you accidentally loaded the wrong host environment and didn't realise some test cases are only available in some environments.

It would be much clearer if all the entries were present all the time, but if you picked a type that can't be found, we have an error message like "The type {name} could not be found. It might not be supported when running on this host environment.". This error could be triggered from `Index.razor`'s `SelectedComponentType` getter, or just inline in the Razor UI logic when `SelectedComponentTypeName` is nonempty but `SelectedComponentType` is null.

Also rather than using assembly qualified names, what stops it from working just having the type name (independently of the assembly)? The logic in `Index.razor`'s `SelectedComponentType` uses `Type.GetType` - doesn't that look at all the loaded assemblies?

File: src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs (hunk 2 above)
Comment (SteveSandersonMS) [reply src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs:14]: Ah I see it's related to `ProtectedBrowserStorageUsageComponent` being in the server project, however I'm still unsure of the necessity as per question https://github.com/dotnet/aspnetcore/pull/25557/files#r482820977.

File: src/Components/test/testassets/BasicTestApp/Index.razor (hunk 3 above)
Comment (pranavkm) [reply src/Components/test/testassets/BasicTestApp/Index.razor:75]: >  The logic in Index.razor's SelectedComponentType uses Type.GetType - doesn't that look at all the loaded assemblies?

It only works if the type is in the [executing assembly](https://docs.microsoft.com/en-us/dotnet/api/system.type.gettype?view=netcore-3.1#System_Type_GetType_System_String_) which requires the additional work around fully qualifying it. I have this tweaked a bit where we rely on the runtime to fail loudly if the type is unable to be found. Perhaps that would suffice here?

File: src/Components/test/testassets/TestServer/PrerenderedStartup.cs [hunk 4]
```diff
@@ -23,7 +23,7 @@ public void ConfigureServices(IServiceCollection services)
             services.AddMvc();
             services.AddServerSideBlazor();
             services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme).AddCookie();
-            services.AddSingleton<LazyAssemblyLoader>();
+            services.AddScoped<LazyAssemblyLoader>();
```
Comment (pranavkm) [inline src/Components/test/testassets/TestServer/PrerenderedStartup.cs:26]: @captainsafia I had trouble running the TestServer app without changing this. `AddServerSideBlazor` adds a scoped `IJSRuntime` instance. Our DI system will complain if a singleton service (in this case `LazyAssemblyLoader`) references a scoped service (`IJSRuntime`). Do you think there's a regression as a result of my changes?

File: src/Components/Server/src/DependencyInjection/ComponentServiceCollectionExtensions.cs (hunk 1 above)
Comment (pranavkm) [reply src/Components/Server/src/DependencyInjection/ComponentServiceCollectionExtensions.cs:38]: > but it does mean they are making more of a conscious choice to plug it into their application.

* Users are making that choice by using these types. Having the service registration isn't a significant cost and we have precedence with MVC registering a whole bunch of optional features \ services in DI behind a single gesture. 

* Previously using these APIs required bringing in an optional dependency (DataProtection), which is no longer the case. I would have felt more strongly about a separate API if this required bringing in more optional services.

* Generally our service registration chain in dependencies. In this case, `AddProtectedBrowserStorage` would only make sense if you had already called `AddServerSideBlazor`. That would require us to either call  `AddServerSideBlazor` on your behalf or use a builder pattern, both of which seem unnecessary.

Comment (javiercn) [reply src/Components/Server/src/DependencyInjection/ComponentServiceCollectionExtensions.cs:38]: The difference is that libraries can start using this without the app developer knowing. I'm not sure if that's something we want, given that this will put data in the browser (even though is protected).

I can imagine it being hard to detect if an app brings in a library that starts making use of this "liberally".

Comment (SteveSandersonMS) [reply src/Components/Server/src/DependencyInjection/ComponentServiceCollectionExtensions.cs:38]: OK, I see the arguments both ways and am fine with either. I think the developer experience is OK either way.

@pranavkm @javiercn If the two of you can come to an agreement on it let's go with that.

File: src/Components/test/testassets/BasicTestApp/Index.razor (hunk 3 above)
Comment (SteveSandersonMS) [reply src/Components/test/testassets/BasicTestApp/Index.razor]: Ah yes, makes sense.

The improvement you've added looks good to me. It's test code - we can always try to streamline it later if we want.

**Review by SteveSandersonMS — APPROVED** [review]
I know there are still a couple of open questions but I'm fine with what's here and whatever decisions you take on them.

File: src/Components/Server/src/DependencyInjection/ComponentServiceCollectionExtensions.cs (hunk 1 above)
Comment (javiercn) [reply src/Components/Server/src/DependencyInjection/ComponentServiceCollectionExtensions.cs:38]: I'll let @pranavkm decide. I don't feel strongly either.

**Comment by Pilchie** [comment]
Does this have a breaking change impact for any customers that have started using it already?

**Comment by pranavkm** [comment]
@Pilchie the would have to remove the package reference and change namespaces when upgrading from rc1. It will be useful to cover some of these packaging changes (there's one more component that's affected) as an announcement.

**Comment by Pilchie** [comment]
Approved for .NET 5 RC2 pending CI checks, but we will need a breaking change announcement for this.

**Comment by pranavkm** [comment]
Announcement filed: https://github.com/aspnet/Announcements/issues/436
